src/VisualStudio/Core/Def/Implementation/LanguageClient/InProcLanguageServer.cs

```diff
@@ -226,6 +230,34 @@ public Task<SymbolInformation[]> GetWorkspaceSymbolsAsync(WorkspaceSymbolParams
             => _requestHandlerProvider.ExecuteRequestAsync<GetTextDocumentWithContextParams, ActiveProjectContexts?>(MSLSPMethods.ProjectContextsName,
                 textDocumentWithContextParams, _clientCapabilities, _clientName, cancellationToken);
 
+        [JsonRpcMethod(SemanticTokensMethods.TextDocumentSemanticTokensName, UseSingleObjectParameterDeserialization = true)]
+        public async Task<SemanticTokens> GetTextDocumentSemanticTokensAsync(SemanticTokensParams semanticTokensParams, CancellationToken cancellationToken)
+        {
+            var tokens = await _requestHandlerProvider.ExecuteSemanticTokensRequestAsync<SemanticTokensParams, SemanticTokens>(
+                SemanticTokensMethods.TextDocumentSemanticTokensName, semanticTokensParams, _tokensCache, _clientCapabilities,
+                _clientName, cancellationToken).ConfigureAwait(false);
+
+            _tokensCache.UpdateCache(semanticTokensParams.TextDocument, tokens);
```

Comment:
>I would guess that the client would wait for edit requests to finish before sending more, but I'm not sure that is true for full document or range.

VS de facto only has one edits request in flight per document at a time, though if you have multiple documents open, we may have multiple out. That said, AFAIK, there's no explicit guarantee of this in the protocol spec.

>cc @davidwengier as well as, this seems like almost a 'write' request because we're updating the server cache.

IMO, the right way to handle this is to treat the cache as just a cache of the _n_ most recent items.. semantic tokens requests should be parallelizable but if there happens to be a cached version that matches, we return that. Then it's just a matter of properly synchronizing via atomic updates or a lock. Though VS doesn't do this.. there's no reason you couldn't have a client that might ask for changes on multiple versions right after one another, and in that case, a cache of _n_ items might be a useful way to think about it.. because scaling to that case just involves increasing the size of the cache.

You could choose to synchronize accesses to this cache by not respond to any other requests until this one completes, but this really shouldn't be a state-impacting request.

Once you have established the snapshot you are on, there's no reason not to start dispatching other requests. Such blocking behavior might significantly decrease throughput.

Reply:
I think Christian's point makes sense - the cache should handle a set of most recent values that we can drop as we get newer ones.  Then we just need to make sure we can synchronize updates to the cache properly.  Even if we handle the requests out of order slightly we still have both results in the cache which can be used by subsequent requests.